### PR TITLE
EIP 747: wallet_watchAsset

### LINKED ISSUE #747: New Web3 Provider Method: eth_watchToken
Today, most web3-compatible browsers have a notion of watching tokens, usually to help a user watch their current balances of tokens.

While it is possible to derive an account's token balances by walking the full history of the blockchain, this is especially impractical in the context of light clients, and so users are frequently instructed to "add this token to your wallet", to help their wallet identify tokens they care about.

To help ease this experience, I'd like to propose a new optional method to be exposed in web3-compatible browsers, `eth_watchToken`.

This method would take a single parameter, a JavaScript object with the following keys:

- address (required): The address of the token to watch, as a hex-prefixed string.
- symbol: A string representing the ticker symbol. Should be three characters. (Required if the token at the given address does not implement a public `symbol() returns string` method.
- decimals: number, representing the number of digits to treat as decimals. (Required if the token at the given address does not implement a public `decimals() returns uint8` method.
- name: string, representing the token's name. Optional.

Web3 browsers should then show some confirmation to the user, requesting their permission to track the given token.

If one of the options (symbol, decimals, name) differs from the result of the token's own public method, I'm not sure what's best, probably default to the page? Open to opinions here, for now this is undefined behavior for the proposed spec.

### Valid error messages:

Valid errors I'd like to propose, for the sake of cross-client consistency, are:

- `Problem prompting the user.`
- `Problem tracking the token.`
- `User rejected request.`

### PR DESCRIPTION
**Pull request by danfinlay**
A method for web3 wallets to allow Dapps to suggest assets that the user might want to "watch". The first reference example is for adding a token to the user's list of watched tokens.

This feature is in [MetaMask v4.10.0](https://github.com/MetaMask/metamask-extension/releases/tag/v4.10.0) and can be tested against [this web app](https://metamask.github.io/Add-Token/#add?tokenAddress=0xdc1adf1d0bf59ec597e1b1b8e3dceb31aafd793f&tokenImage=https%3A%2F%2Fpbs.twimg.com%2Fprofile_images%2F723268858644037632%2Fu-kjO4___400x400.jpg&tokenName=Danbucks&tokenNet=3&tokenSymbol=DBX&tokenDecimals=2).

Latest discussion on the feature can be found here:
https://ethereum-magicians.org/t/eip-747-eth-watchtoken/1048

Fixes #747.

### CODE REVIEW

**Comment by Arachnid** [comment]
Where does 747 come from?

**Comment by bitpshr** [comment]
What's the status of this? Does any formatting need to be modified or information clarified for this to achieve "draft" status? I believe both MetaMask and Gnosis have implemented this feature.

cc @Arachnid @danfinlay

**Comment by phraktle** [comment]
@danfinlay there's a UX problem that this spec does not adequately address: when the user has already added the asset (manually or through `wallet_watchAsset`), there should be a way to avoid asking them to add it repeatedly in a dApp.

IMO, this would be best achieved by specifying that a `wallet_watchAsset` should not prompt the user in case the asset is already monitored by the wallet, instead silently reporting success. This way a dApp could detect that the user has some tokens of a specific type and prompt the user to add it, which would just be a no-op in case they have already added these.

**Comment by phraktle** [comment]
@estebanmino the `watchAsset` API is inadequate even for the purpose it was meant to address. See https://github.com/ethereum/EIPs/pull/1426#issuecomment-472350298

Until Ethereum and Metamask can properly address the problem of detecting tokens in wallets, I urge Metamask maintainers to merge requests to add tokens that fulfill the listed criteria (such as AEUR), instead of further politicizing the process and making it a popularity contest).

**Comment by axic** [comment]
> Where does 747 come from?

@Arachnid Apparently the issue number #747 where this was discussed.

File: EIPS/eip-747.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
+---
+eip: <to be assigned>
```
Comment (axic) [inline EIPS/eip-747.md]: Please use 747.

File: EIPS/eip-747.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+title: Add wallet_watchAsset to Provider
+author: Dan Finlay (@danfinlay), Esteban Mino (@estebanmino)
+discussions-to: https://ethereum-magicians.org/t/eip-747-eth-watchtoken/1048
+status: WIP
```
Comment (axic) [inline EIPS/eip-747.md]: Please use `Draft`.

File: EIPS/eip-747.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+type: Interface
+category: Interface
+created: August 13, 2018
```
Comment (axic) [inline EIPS/eip-747.md]: Please use `2018-08-13`.

**Comment by axic** [comment]
Since this ERC references the "provider api", does this depend on http://eips.ethereum.org/EIPS/eip-1193 ?

**Comment by axic** [comment]
@danfinlay can you please respond to the questions above?

**Comment by danfinlay** [comment]
> Since this ERC references the "provider api", does this depend on http://eips.ethereum.org/EIPS/eip-1193 ?

No, it does not. That EIP is an extension of the provider, and this EIP deals with a wallet-namespaced method within the JSON-RPC message layer, which is a lower level than that proposal.

I've implemented the other change requests.

**Comment by axic** [comment]
> That EIP is an extension of the provider, and this EIP deals with a wallet-namespaced method within the JSON-RPC message layer, which is a lower level than that proposal.

Oh, then it should depend on 1474.

**Comment by danfinlay** [comment]
Updated.

File: EIPS/eip-747.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
+---
+eip: 747
+title: Add wallet_watchAsset to Provider
+author: Dan Finlay (@danfinlay), Esteban Mino (@estebanmino)
+discussions-to: https://ethereum-magicians.org/t/eip-747-eth-watchtoken/1048
+status: Draft
+type: Interface
```
Comment (axic) [inline EIPS/eip-747.md]: Dunno why the validator hasn't caught this, but it should be: `type: Standards Track`

Comment (danfinlay) [reply EIPS/eip-747.md]: Corrected.

**Comment by axic** [comment]
> EIPS/eip-747.md:40: specifing  ==> specifying

@danfinlay sorry this is still failing validation due the above typo.

**Comment by danfinlay** [comment]
Fixed.

**Comment by axic** [comment]
:clap:

**Comment by danfinlay** [comment]
😅👍

File: EIPS/eip-747.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+		address,
+		symbol,
+		decimals,
+		image: 'linktoimage.jpg'
+	}
+})
+```
+Upon calling this request, the user should be prompted with the opportunity to add this token to their wallet:
+
+![ezgif com-video-to-gif](https://user-images.githubusercontent.com/12115171/44558730-2666e300-a71c-11e8-9098-e27ec99b88fb.gif)
```
Comment (axic) [inline EIPS/eip-747.md]: Damn, didn't notice these embedded gif links.

@danfinlay can you move them into `assets/eip-747`?

Comment (danfinlay) [reply EIPS/eip-747.md]: Sorry about that: https://github.com/ethereum/EIPs/pull/2150

**Pull request merged**